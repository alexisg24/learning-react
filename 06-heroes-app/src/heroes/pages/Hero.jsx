import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroByID } from '../helpers'
import { useMemo } from 'react'

const HeroPage = () => {
  const { id } = useParams()
  const navigate = useNavigate()
  const hero = useMemo(() => getHeroByID(id), [id])
  const url = useMemo(() => `/assets/heroes/${id}.jpg`, [id])
  const handleNavitageBack = () => {
    return navigate(-1)
  }

  if (!hero) {
    return (<Navigate to='/marvel' />)
  }
  return (
    <section className='row mt-5 animate__animated animate__fadeInLeftBig'>
      <div className='col-4'>
        <img
          src={url}
          alt={hero.superhero}
          className='img-thumbnail'
        />
      </div>
      <div className='col-8'>
        <h3>{hero.superhero}</h3>
        <ul className='list-group list-group-flush'>
          <li className='list-group-item'><b>Alter ego:</b> {hero.alter_ego}</li>
          <li className='list-group-item'><b>Publisher:</b> {hero.publisher}</li>
          <li className='list-group-item'><b>First Appearance:</b> {hero.first_appearance}</li>
        </ul>

        {
        (hero.characters !== hero.alter_ego) &&
        (
          <>
            <h5 className='mt-3'>Characters</h5>
            <p>{hero.characters}</p>
          </>
        )
        }
        <button
          className='btn btn-outline-primary'
          onClick={handleNavitageBack}
        >Back
        </button>
      </div>
    </section>
  )
}

export { HeroPage }
