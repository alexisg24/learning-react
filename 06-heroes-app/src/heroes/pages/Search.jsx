import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../components/HeroCard'
import { getHeroesByName } from '../helpers'

const SearchPage = () => {
  const navitage = useNavigate()
  const { search } = useLocation()
  const { hero = '' } = queryString.parse(search)
  const { searchText, onInputChange } = useForm({ searchText: hero })
  const heroes = getHeroesByName(hero)

  const onSearchSubmit = (event) => {
    event.preventDefault()
    if (searchText.trim().length <= 1) return
    navitage(`?hero=${searchText.toLowerCase().trim()}`, { replace: true })
  }

  return (
    <>
      <h1>SearchPage</h1>
      <hr />
      <div className='row'>
        <section className='col-5'>
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              type='text'
              placeholder='Search a hero'
              className='form-control'
              name='searchText'
              autoComplete='off'
              value={searchText}
              onChange={onInputChange}
            />
            <button className='btn btn-outline-primary mt-1'>Search</button>
          </form>
        </section>
        <section className='col-7'>
          <h4>Result</h4>
          <hr />
          {
            (hero === '')
              ? (
                <article className='alert alert-primary'>
                  Search a Hero
                </article>
                )
              : (heroes.length === 0) && (
                <article className='alert alert-danger'>
                  No Hero with <b>{hero}</b>
                </article>
                )
          }

          {
            heroes.map(hero => (<HeroCard key={hero.id} {...hero} />))
          }
        </section>
      </div>
    </>
  )
}

export { SearchPage }
