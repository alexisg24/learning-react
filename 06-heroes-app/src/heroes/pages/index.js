export * from './DcPage'
export * from './Marvel'
export * from './Hero'
export * from './Search'
