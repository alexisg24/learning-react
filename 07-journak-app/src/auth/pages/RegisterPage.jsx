import { Link as RouterLink } from 'react-router-dom'
import { Google } from '@mui/icons-material'
import { Button, Grid, Link, TextField, Typography } from '@mui/material'
import { AuthLayout } from '../layout/AuthLayout'

const RegisterPage = () => {
  return (
    <AuthLayout title='Register'>
      <form>
        <Grid container>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField label='Full Name' type='text' placeholder='Jonh Doe' fullWidth />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField label='Email' type='email' placeholder='mail@google.com' fullWidth />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField label='Password' type='password' placeholder='Password' fullWidth />
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            <Grid item xs={12}>
              <Button variant='contained' fullWidth>Create Account</Button>
            </Grid>
          </Grid>
        </Grid>

        <Grid container direction='row' justifyContent='end'>
          <Typography sx={{ mr: 1 }}>Do you have an account?</Typography>
          <Link component={RouterLink} color='inherit' to='/auth/login'>
            Login
          </Link>
        </Grid>
      </form>
    </AuthLayout>
  )
}

export { RegisterPage }
