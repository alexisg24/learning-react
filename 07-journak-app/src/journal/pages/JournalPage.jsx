import { MailOutline } from '@mui/icons-material'
import { Typography } from '@mui/material'

const JournalPage = () => {
  return (
    <>
      <Typography variant='h1'>Journal Page</Typography>
      <MailOutline />
    </>
  )
}

export { JournalPage }
