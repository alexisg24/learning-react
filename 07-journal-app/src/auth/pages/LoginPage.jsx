import { Link as RouterLink } from 'react-router-dom'
import { Google } from '@mui/icons-material'
import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material'
import { AuthLayout } from '../layout/AuthLayout'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux'
import { startEmailAndPasswordSignIn, startGoogleSignIn } from '../../store/auth'
import { useMemo } from 'react'

const LoginPage = () => {
  const { status, errorMessage } = useSelector(state => state.auth)
  const dispatch = useDispatch()

  const { email, password, onInputChange } = useForm({
    email: 'john@doe.com',
    password: '123456'
  })

  const isAuthenticating = useMemo(() => status === 'checking', [status])

  const onSubmit = (event) => {
    event.preventDefault()
    dispatch(startEmailAndPasswordSignIn({ email, password }))
  }

  const onGoogleSignIn = (event) => {
    console.log('Google sign in')
    dispatch(startGoogleSignIn())
  }

  return (
    <AuthLayout title='Login'>
      <form onSubmit={onSubmit}>
        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField label='Email' type='email' placeholder='mail@google.com' fullWidth onChange={onInputChange} value={email} name='email' />
          </Grid>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField label='Password' type='password' placeholder='Password' fullWidth onChange={onInputChange} value={password} name='password' />
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            <Grid item xs={12} display={errorMessage ? '' : 'none'}>
              <Alert severity='error'>{errorMessage}</Alert>
            </Grid>
            <Grid item xs={12} sm={6}>
              <Button variant='contained' fullWidth type='submit' disabled={isAuthenticating}>Login</Button>
            </Grid>
            <Grid item xs={12} sm={6}>
              <Button variant='contained' fullWidth onClick={onGoogleSignIn} disabled={isAuthenticating}>
                <Google />
                <Typography sx={{ ml: 1 }}>Google</Typography>
              </Button>
            </Grid>
          </Grid>
        </Grid>

        <Grid container direction='row' justifyContent='end'>
          <Link component={RouterLink} color='inherit' to='/auth/register'>
            Create an Account
          </Link>
        </Grid>
      </form>
    </AuthLayout>
  )
}

export { LoginPage }
