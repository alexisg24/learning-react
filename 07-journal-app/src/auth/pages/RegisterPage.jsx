import { Link as RouterLink } from 'react-router-dom'
import { Button, Grid, Link, TextField, Typography } from '@mui/material'
import { AuthLayout } from '../layout/AuthLayout'
import { useForm } from '../../hooks/useForm'
import { useState } from 'react'
import { startCreatingUserWithEmailPassword } from '../../store/auth'
import { useDispatch } from 'react-redux'
const formData = {
  displayName: '',
  email: '',
  password: ''
}

const formValidation = {
  displayName: [(value) => value.length >= 1, 'Name is required'],
  email: [(value) => value.includes('@'), 'Email needs to be valid'],
  password: [(value) => value.length >= 6, 'Password is required and must be at least 6 characters']
}
const RegisterPage = () => {
  const dispatch = useDispatch()
  const [formSubmited, setFormSubmited] = useState(false)
  const {
    displayName,
    email,
    password,
    onInputChange,
    isFormValid,
    displayNameValid,
    emailValid,
    passwordValid
  } = useForm(formData, formValidation)

  const onSubmit = (event) => {
    event.preventDefault()
    setFormSubmited(true)
    if (!isFormValid) return true
    dispatch(startCreatingUserWithEmailPassword({ email, password, displayName }))
  }

  return (
    <AuthLayout title='Register'>
      <h1>FormValid {isFormValid ? 'true' : 'false'}</h1>
      <form onSubmit={onSubmit}>
        <Grid container>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label='Full Name'
              type='text'
              placeholder='Jonh Doe'
              fullWidth onChange={onInputChange}
              value={displayName}
              name='displayName'
              error={!!displayNameValid && formSubmited}
              helperText={displayNameValid}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label='Email'
              type='email'
              placeholder='mail@google.com'
              fullWidth onChange={onInputChange}
              value={email}
              name='email'
              error={!!emailValid && formSubmited}
              helperText={emailValid}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label='Password'
              type='password'
              placeholder='Password'
              fullWidth
              onChange={onInputChange}
              value={password}
              name='password'
              error={!!passwordValid && formSubmited}
              helperText={passwordValid}
            />
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            <Grid item xs={12}>
              <Button variant='contained' fullWidth type='submit'>Create Account</Button>
            </Grid>
          </Grid>
        </Grid>

        <Grid container direction='row' justifyContent='end'>
          <Typography sx={{ mr: 1 }}>Do you have an account?</Typography>
          <Link component={RouterLink} color='inherit' to='/auth/login'>
            Login
          </Link>
        </Grid>
      </form>
    </AuthLayout>
  )
}

export { RegisterPage }
