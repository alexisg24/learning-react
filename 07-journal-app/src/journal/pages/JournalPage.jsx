import { IconButton } from '@mui/material'
import { JournalLayout } from '../layout/JournalLayout'
import { NoteView, NothingSelected } from '../views'
import { AddOutlined } from '@mui/icons-material'
import { useDispatch, useSelector } from 'react-redux'
import { startNewNote } from '../../store/journal'

const JournalPage = () => {
  const dispatch = useDispatch()
  const { isSaving, active: activeNote } = useSelector(state => state.journal)

  const onClickNewNote = () => {
    dispatch(startNewNote())
  }
  return (
    <JournalLayout>
      {(activeNote) ? <NoteView /> : <NothingSelected />}
      <IconButton
        size='large'
        sx={{
          color: 'white',
          backgroundColor: 'error.main',
          ':hover': { backgroundColor: 'error.main', opacity: 0.6, transition: 'all .3s ease-out' },
          position: 'fixed',
          right: 50,
          bottom: 50
        }}
        disabled={isSaving}
        onClick={onClickNewNote}
      >
        <AddOutlined />
      </IconButton>
    </JournalLayout>
  )
}

export { JournalPage }
