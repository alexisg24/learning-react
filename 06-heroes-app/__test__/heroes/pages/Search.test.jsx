import { fireEvent, render, screen } from '@testing-library/react'
import { SearchPage } from '../../../src/heroes/pages/Search'
import { MemoryRouter } from 'react-router-dom'

const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate
}))

describe('tests in Search', () => {
  beforeAll(jest.resetAllMocks)
  test('should return the component with default values', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })

  test('should return Batman and input of queryString', () => {
    const query = 'batman'
    render(
      <MemoryRouter initialEntries={[`/search?hero=${query}`]}>
        <SearchPage />
      </MemoryRouter>
    )
    const inputValue = screen.getByRole('textbox')
    expect(inputValue.value).toBe(query)

    const heroImg = screen.getByRole('img')
    expect(heroImg.src).toContain(query)

    expect(() => screen.getByLabelText('search-hero').toThrow()).toBeTruthy()
    expect(() => screen.getByLabelText('hero-not-found').toThrow()).toBeTruthy()
  })

  test('should return error if hero not found', () => {
    const query = '2313321'
    render(
      <MemoryRouter initialEntries={[`/search?hero=${query}`]}>
        <SearchPage />
      </MemoryRouter>
    )
    expect(screen.getByLabelText('hero-not-found')).toBeTruthy()
  })

  test('should call navigate to new screen', () => {
    const query = 'batman'
    render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )
    const inputElement = screen.getByRole('textbox')
    fireEvent.input(inputElement, { target: { value: query } })
    const formButton = screen.getByRole('button')
    fireEvent.click(formButton)
    expect(mockedUseNavigate).toHaveBeenCalledWith(`?hero=${query}`, { replace: true })
  })
})
