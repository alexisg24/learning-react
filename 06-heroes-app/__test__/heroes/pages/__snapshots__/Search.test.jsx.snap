// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in Search should return the component with default values 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <hr />
  <div
    class="row"
  >
    <section
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </section>
    <section
      class="col-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <article
        aria-label="search-hero"
        class="alert alert-primary"
      >
        Search a Hero
      </article>
    </section>
  </div>
</div>
`;
